Fabric.Identity.API/Services/ClaimsService.cs

```diff
@@ -0,0 +1,147 @@
+                }
+            }
+
+            //remove the user id claim from the claims collection and move to the userId property
+            //also set the name of the external authentication provider
+            claims.Remove(userIdClaim);
+
+            //get the client id from the auth context
+            result.AdditionalClaims = this.GenerateAdditionalClaims(claims);
+            result.AuthenticationProperties = this.GenerateAuthenticationProperties(info);
+
+            return result;
+        }
+
+        public string GetSubjectId(List<Claim> claims, User user, string schemaItem)
+        {
+            string subjectId = null;
+            if (this.IsAzureEnabled(schemaItem))
+            {
+                subjectId = claims.FirstOrDefault(x => x.Type == AzureActiveDirectoryJwtClaimTypes.OID || x.Type == AzureActiveDirectoryJwtClaimTypes.OID_Alternative)
+                                  .Value;
+            }
+
+            if(subjectId == null)
+            {
+                subjectId = user.SubjectId;
+            }
+
+            return subjectId;
+        }
+
+        private Claim[] GenerateAdditionalClaims(List<Claim> previousClaims)
+        {
+            var additionalClaims = new List<Claim>();
+            //if the external system sent a session id claim, copy it over
+            var sid = previousClaims.FirstOrDefault(x => x.Type == JwtClaimTypes.SessionId);
+            if (sid != null)
+            {
+                additionalClaims.Add(new Claim(JwtClaimTypes.SessionId, sid.Value));
+            }
+
+            //if the external provider issues groups claims, copy it over
+            var groupClaims = previousClaims.Where(c => c.Type == "groups").ToList();
+            if (groupClaims.Any())
+            {
+                additionalClaims.AddRange(groupClaims);
+            }
+
+            return additionalClaims.ToArray();
+        }
+
+        private AuthenticationProperties GenerateAuthenticationProperties(AuthenticateInfo info)
+        {
+            //if the external provider issued an id_token, we'll keep it for signout
+            AuthenticationProperties props = null;
+            var id_token = info.Properties.GetTokenValue("id_token");
+            if (id_token != null)
+            {
+                props = new AuthenticationProperties();
+                props.StoreTokens(new[] { new AuthenticationToken { Name = "id_token", Value = id_token } });
+            }
+
+            return props;
+        }
+
+        private bool ValidateAzureAD(List<Claim> claims)
+        {
+            var issuerClaim = claims.FirstOrDefault(x => x.Type == JwtClaimTypes.Issuer);
+            if (issuerClaim == null)
+            {
+                throw new MissingIssuerClaimException(ExceptionMessageResources.MissingIssuerClaimMessage);
+            }
+
+            return !this._appConfiguration.AzureActiveDirectorySettings.IssuerWhiteList.Contains(issuerClaim.Issuer);

```

Comment:
so the validate function returns true if the issuer is NOT in the whitelist?    
Would it be more clear to call the function IsAzureADInvalid ? 

or return true if it's in the whitelist and change the logic that uses it.

Reply:
yeah the ! needs to be moved.